Add the first version of copilot instructions (#3238)
Azure/autorest.typescript: .github/copilot-instructions.md

# GitHub Copilot Instructions for autorest.typescript

> This repository focuses on the TypeSpec TypeScript generator, which generates TypeScript client libraries from [TypeSpec](https://typespec.io/) specifications. TypeSpec is the recommended approach for describing cloud service APIs and generating client libraries.

Author: removed.

> > **Note:** While this repository also contains the AutoRest TypeScript generator for OpenAPI/Swagger specifications, we recommend using TypeSpec for new projects. The AutoRest generator is in maintenance mode and will be deprecated in the future.

Reviewer: Also this might not be very helpful. I Think a note on the projects that are actually in maintenance mode stating, this is in maintenance mode, should ignore unless explicitly asked to use or edit

## Project Structure

- `packages/typespec-ts/` - TypeSpec TypeScript generator

>   - Generates TypeScript clients from TypeSpec definitions

Reviewer: ```suggestion
  - Generates TypeScript http clients from TypeSpec definitions
```

>   - Supports both RLC and Modular SDK generation

Reviewer: We should make sure to include a description on what RLC is, otherwise this could cause copilot to hallucinate

Author: sure.

  - RLC provides low-level REST client operations
  - Modular SDK offers modern, enhanced developer experience
  - Uses TypeSpec compiler for processing

- `packages/rlc-common/` - Common utilities for REST Level Client generation

  - Core RLC patterns and utilities

>   - Shared code for TypeSpec RLC implementation

Reviewer: maybe remove this

>   - Location for RLC changes and improvements

Author: The unclear part is rlc-common would also include some utils which Modular would use.

- `packages/typespec-test/` - Test utilities and fixtures for TypeSpec code generation

  - Test framework for TypeSpec generator
  - Sample TypeSpec definitions
  - Validation utilities
  - Test fixtures and helpers

- `packages/autorest.typescript/` - AutoRest TypeScript generator

>   > **Note:** In maintenance mode, will be deprecated

Reviewer: ```suggestion
  > **Note:** @autorest/typescript is in maintenance mode and should not be used as reference nor edited unless explicitly requested.
```

## AutoRest TypeScript Generator

> The AutoRest TypeScript generator (`autorest.typescript`) transforms OpenAPI/Swagger specifications into TypeScript client libraries. It supports two client styles:

Reviewer: Maybe remove all other text under this section and just keep the suggestion to avoid overloading and confusing copilot
```suggestion
The AutoRest TypeScript generator (`autorest.typescript`)  is in maintenance mode and should not be used as reference nor edited unless explicitly requested. For more information [AutoRest documentation](https://github.com/Azure/autorest).
```

1. **REST Level Client (RLC)** - Low-level mapping to REST operations
2. **High Level Client (HLC)** - Higher-level abstraction layer

For detailed documentation on the AutoRest generator, please refer to the [AutoRest documentation](https://github.com/Azure/autorest).

## TypeSpec Test Framework

The TypeSpec test framework (`typespec-test`)provides utilities and test fixtures for validating TypeSpec code generation. Here are the key aspects to consider:

> ## TypeSpec TypeScript Generator

Reviewer: I wonder if we should use Emitter consistently to make it easier for Copilot to reason about, Emitter is the term used in TypeSpec docs

The TypeSpec TypeScript generator (`typespec-ts`) generates TypeScript client libraries from [TypeSpec](https://typespec.io/) specifications. TypeSpec is a language specifically designed for API development that:

> - Provides a more intuitive way to describe APIs

Reviewer: I don't think we need to provide this description about TypeSpec, sharing a reference link as above should be enough. Unless we are pointing out something specific I don't think these summary descriptions of external things like TypeSpec are of much value to copilot

- Supports rich type systems and inheritance
- Enables better code organization through namespaces and imports
- Allows custom decorators for metadata and behavior
- Generates consistent API artifacts (clients, documentation, etc.)

The generator supports two distinct SDK styles:

### SDK Generation Styles

> 1. **REST Level Client (RLC)**

Reviewer: We talk about RCL and Modular at te beginning. I wonder if it would be better if we added a Key concepts section at the beginning where we describe these. Then here we just say it can generate: - Rlc libraries - Modular Libraries

   - Low-level, direct mapping to REST API operations
   - One-to-one correspondence with API endpoints
   - Follows REST Level Client patterns
   - Minimal abstraction over HTTP operations
   - Ideal for custom client implementations

2. **Modular SDK**

   - Next generation of High Level Client (HLC)
   - Modern, enhanced developer experience
   - Structured around service capabilities
   - Strong type safety and IntelliSense support
   - Improved error handling and diagnostics

> ## Upgrade TypeSpec relevant dependencies for TypeSpec TypeScript generator

Reviewer: ```suggestion
## How to Upgrade TypeSpec dependencies for @azure-tools/typespec-ts (packagest/typespec-ts)
```

> To upgrade dependencies for the TypeSpec TypeScript generator, we need to focus on the `packages/typespec-ts/` and `packages/typespec-test/` packages only, not `packages/rlc-common` or `packages/autorest.typescript`.

Reviewer: ```suggestion
When upgrading TypeSpec dependencies only work on `packages/typespec-ts/` and `packages/typespec-test/` . `packages/rlc-common` and `packages/autorest.typescript` should not be edited.
```

> - First look at the `package.json` files in these directories to identify the dependencies including devDependencies, peerDependencies, and dependencies.

Reviewer: ```suggestion
- Identify dependencies within the @typespec or @azure-tools npm-scopes under "dependencies", "devDependencies" and "peerDependencies" in package.json
- You should use "next" tag for dependencies on `@typespec/http-specs`, `@typespec/spector`, `@azure-tools/azure-http-specs`, and `@typespec/spec-api`
- You should use "latest" for the other TypeSpec dependencies
- Run "rush update" after editing the dependencies in package.json
```

- Do not update dependency for @typespec/ts-http-runtime.
- For spector relevant dependencies, the latest version which is tagged with `next` on npm including `@typespec/http-specs`, `@typespec/spector`, `@azure-tools/azure-http-specs`, and `@typespec/spec-api`.
- For other dependencies, check the latest versions which is tagged with `latest` on npm and update them accordingly.
- After updating the versions, run `rush update` to ensure all dependencies are correctly installed and the lock files are updated.
- Do run `rush build` to build the entire monorepo to check if any building issues introduced by upgrading.
- Do run `rush format` to format the codebase.
- Do run commands under `packages/typespec-ts/` to work on the TypeSpec TypeScript generator.
- Do run `rushx test` to run the testing and follow the instructions below to fix any test failures.
- Do run `rushx lint` to run the linter.
- Do run commands under `packages/typespec-test/` to work on the TypeSpec test framework.
- Do run `rushx test` to run the TypeSpec test framework tests to regen smoke testing for typespec and if any issue try to fix it.

## How to run and fix test failures in TypeSpec TypeScript generator

> Please ensure you have ran `rush update` and `rush build` before running tests. Before you start working on fixing test failures in the TypeSpec TypeScript generator, you need to understand which kinds of test failures you are dealing with. The tests in the TypeSpec TypeScript generator can be broadly categorized into:

Reviewer: ```suggestion
Run `rush update` and `rush build` before running tests. 

The tests in the TypeSpec TypeScript generator can be categorized into:
```
